### PR TITLE
Dahee/openroad

### PR DESCRIPTION
**Pull request by daheehee**
adding openroad interfacing capabilities

### CODE REVIEW

File: openroad_interface/test_files/mm_test.gml [hunk 1]
```diff
@@ -0,0 +1,797 @@
+graph [
```
Comment (fazerlicourice7) [inline openroad_interface/test_files/mm_test.gml:1]: NOw that this is being integrated into the codesign repo, we should have these tests point to `codesign/src/architectures/<filename>.gml` instead of having a copy in here

File: openroad_interface/test/test_ibexsky130hd.tcl [hunk 2]
```diff
@@ -0,0 +1,234 @@
+# ibex sky130hd
```
Comment (fazerlicourice7) [inline openroad_interface/test/test_ibexsky130hd.tcl:1]: are these modified from the defaults at all? is there a reason we have a copy here? 
Or are these the scripts that you've modified to use the custom .def and ignore the .v

File: openroad_interface/parasitic_estimation.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
+from var import directory
+import re
+import math
+
+
+def total_euclidean_distance(net_list, coord_data, unit):
+    the_component = net_list[0]
```
Comment (fazerlicourice7) [inline openroad_interface/parasitic_estimation.py:7]: Can we see if we can extract a `Half perimeter Wire Length` number from open road itself. The fact that our wirelengths are overerstimates is not ok, the HPWL should be an approximation within openroad that lines up with our estimates.

issue with this is that it uses the center of the std cell for the distances, when in reality the pins are all over, so this largely overestimates wirelength

File: openroad_interface/parasitic_estimation.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    result = 0.0
+    for x in range(1, len(net_list)):
+        result += math.sqrt(pow(coord_data[the_component]["x"]/unit - coord_data[net_list[x]]["x"]/unit, 2) + pow(coord_data[the_component]["y"]/unit - coord_data[net_list[x]]["y"]/unit, 2))
+    return result
+
+def parasitic_estimation(lef_pitch, layer_res, layer_cap, units):
```
Comment (fazerlicourice7) [inline openroad_interface/parasitic_estimation.py:13]: Add function docs. What does the function do, what are the inputs, what data type. 
Better yet, add the types in the signature, python3 style, along with a return type

https://docs.python.org/3/library/typing.html

File: openroad_interface/exec.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
+
+def exec(design_file, test_file):
+    from var import directory
+    from def_generator import def_generator
+    import networkx as nx
+    import os
+    import shutil
+    import parasitic_estimation as pe
+    from length_calculations import length_calculations
+    import parasitic_calc as pc
```
Comment (fazerlicourice7) [inline openroad_interface/exec.py:10]: imports should be outside the function

File: openroad_interface/test/test_sky130hd.tcl [hunk 6]
```diff
@@ -0,0 +1,228 @@
+# gcd flow pipe cleaner
+source "helpers.tcl"
+source "flow_helpers.tcl"
+source "sky130hd/sky130hd.vars"
+
+set synth_verilog "gcd_sky130hd.v"
```
Comment (fazerlicourice7) [inline openroad_interface/test/test_sky130hd.tcl:6]: remove?

File: openroad_interface/test_files/Nangate45_tech.lef [hunk 7]
```diff
@@ -0,0 +1,779 @@
+#
```
Comment (fazerlicourice7) [inline openroad_interface/test_files/Nangate45_tech.lef:1]: remove if not modified

File: openroad_interface/parasitic_estimation.py (hunk 4 above)
Comment (fazerlicourice7) [reply openroad_interface/parasitic_estimation.py:13]:
```
def parasitic_estimation(lef_pitch: float, layer_res: float, layer_cap: float, units: float) -> Dict():
    '''
        Calculate approximate R, C, wirelength for nets.

params: 
    lef_pitch: ....
    layer_res: .....

    '''

File: openroad_interface/parasitic_calc.py [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+                    new_list[ref_pin_2_index][pin_index_2] = ""
+                    new_list[ref_pin_2_index][res_index] = new_res
+                    return res_series(new_list, pin_count)
+                else :
+                    continue
+        return [net_list, pin_count]
+
+    ### read def spef file assign net with capacitance and resistance (total resistance)
+    from functions import find_val_two, find_val_xy, find_val, value, format, clean
+    from var import directory
```
Comment (fazerlicourice7) [inline openroad_interface/parasitic_calc.py:86]: imports at top.

File: openroad_interface/parasitic_calc.py [hunk 9]
```diff
@@ -0,0 +1,139 @@
+
```
Comment (fazerlicourice7) [inline openroad_interface/parasitic_calc.py:1]: first party imports - python builtins

third parrty imports  - pip installed libraries

second party imports  - our libraries.

File: openroad_interface/exec.py [hunk 10]
```diff
@@ -0,0 +1,76 @@
+
+def exec(design_file, test_file):
```
Comment (fazerlicourice7) [inline openroad_interface/exec.py:2]: rename routing_validation

File: openroad_interface/test_files/gcd_nangate45-tcl.spef [hunk 11]
```diff
@@ -0,0 +1,2815 @@
+*SPEF "ieee 1481-1999"
```
Comment (fazerlicourice7) [inline openroad_interface/test_files/gcd_nangate45-tcl.spef:1]: remove

File: openroad_interface/test_files/tech_params.yaml [hunk 12]
```diff
@@ -0,0 +1,890 @@
+area: # nm^2
```
Comment (fazerlicourice7) [inline openroad_interface/test_files/tech_params.yaml:1]: `codesign/src/params/tech_params.yaml`

File: openroad_interface/test/test_ibexsky130hd.tcl (hunk 2 above)
Comment (fazerlicourice7) [reply openroad_interface/test/test_ibexsky130hd.tcl:1]: update dir above to be `tcl`

File: openroad_interface/test_files/Nangate45_stdcell.lef [hunk 13]
```diff
@@ -0,0 +1,12137 @@
+# NangateOpenCellLibrary.macro.mod.lef
```
Comment (fazerlicourice7) [inline openroad_interface/test_files/Nangate45_stdcell.lef:1]: update dir above to be `std_cell_lef`

File: openroad_interface/graph_plotter.py [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+        xtick.append(designs[x])
+    ax.set_xticklabels(xtick) 
+    plt.ylabel(units)
+    plt.title(title)
+
+    if show_flier:
+        plt.savefig(directory + '.jpeg')
+    else:
+        plt.savefig(directory + '-noflier.jpeg')
+    plt.close(fig)
```
Comment (fazerlicourice7) [inline openroad_interface/graph_plotter.py:60]: save in new dir called `figs`

**Review by fazerlicourice7 — COMMENTED** [review]
Add a README.md explaining any setup. fine here to point to openroad setup docs

File: openroad_interface/exec.py [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+    for output_pin in net_out_dict:
+        for pin in output_dict[output_pin]:
+            graph[output_pin][pin]['net'] = net_out_dict[output_pin]
+            graph[output_pin][pin]['net_length'] = length_list[net_out_dict[output_pin]]
+            graph[output_pin][pin]['net_res'] = net_res[net_out_dict[output_pin]][3]
+            graph[output_pin][pin]['net_cap'] = net_cap[net_out_dict[output_pin]]
+        res_graph_data.append(float(net_res[net_out_dict[output_pin]][3]))
+        cap_graph_data.append(float(net_cap[net_out_dict[output_pin]])* pow(10,4))
+        len_graph_data.append(float(length_list[net_out_dict[output_pin]]))
+
```
Comment (fazerlicourice7) [inline openroad_interface/exec.py:43]: All this should be in a separate function that takes a nx.Digraph as input, and returns a separate nx.Digraph with all the edge attributes.

Comment (fazerlicourice7) [reply openroad_interface/exec.py:43]: call the file `place_n_route.py`

File: openroad_interface/exec.py [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+    if not os.path.exists("results/"):
+        os.makedirs("results/")
+    nx.write_gml(graph, "results/final.gml")
+
+    # 5. generating estimations
+    estimation_dict = pe.parasitic_estimation(lef_pitch, layer_res, layer_cap, units)
```
Comment (fazerlicourice7) [inline openroad_interface/exec.py:49]: separate function for the estimates. with the same interface

File: openroad_interface/place_n_route.py [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+        net_graph_data.append(net_out_dict[output_pin])
+        res_graph_data.append(float(net_res[net_out_dict[output_pin]]))
+        cap_graph_data.append(float(net_cap[net_out_dict[output_pin]]) * pow(10,4))
+        len_graph_data.append(float(length_dict[net_out_dict[output_pin]]))
+
+    if not os.path.exists("results/"):
+        os.makedirs("results/")
+    nx.write_gml(graph, "results/openroad_" + design_name + ".gml")
+
+    return {"res": res_graph_data, "cap": cap_graph_data, "length": len_graph_data, "net": net_graph_data}
```
Comment (fazerlicourice7) [inline openroad_interface/place_n_route.py:46]: writing the graph is fine, but I also want you to return it. 
Where this will be called in the codesign framework, we need the graph object back.

File: openroad_interface/estimation.py [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+
+    return:
+        dict: all res cap length data
+    '''
+    final_def_file = directory + "results/final_generated-tcl.def"
+    pattern = r"_\w+_\s+\w+\s+\+\s+PLACED\s+\(\s*\d+\s+\d+\s*\)\s+\w+\s*;"
+    net_pattern =  r'-\s(_\d+_)\s((?:\(\s_\d+_\s\w+\s\)\s*)+).*'
+    component_pattern = r'(_\w+_)'
+
+    units = lef_data["units"]
```
Comment (fazerlicourice7) [inline openroad_interface/estimation.py:57]: The only information we're using from the def file here is the placement location right?
For the estimation we shouldn't be using anything else from openroad

File: openroad_interface/estimation.py [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+    wire_global_data = open(wire_global_file)
+    wire_global_lines = wire_global_data.readlines()
+    for line in wire_global_lines:
+        if re.search(global_pattern, line) != None:
+            match = re.search(global_pattern, line)
+            globa_length = match.group(0).split()[2]
+            global_length_data.append(float(globa_length))
+    return global_length_data
+
+def parasitic_estimation(graph, design_name: str, net_out_dict: dict, node_output: dict, lef_data: dict) -> dict:
```
Comment (fazerlicourice7) [inline openroad_interface/estimation.py:38]: Let's split this function up into two. 
One which extracts the positions from the def files and adds them as node attributes to the graph.
And then this function which takes the graph with the position attributes as an input parameter.

Comment (fazerlicourice7) [reply openroad_interface/estimation.py:38]: I'll explain rationale for this in person

File: openroad_interface/place_n_route.py [hunk 20]
```diff
@@ -0,0 +1,48 @@
+import os 
+
+import networkx as nx
+
+from var import directory
+import length_calculations as lc
+import parasitic_calc as pc
+
+def place_n_route(graph, design_name: str, net_out_dict: dict, node_output: dict, lef_data: dict) -> dict:
```
Comment (fazerlicourice7) [inline openroad_interface/place_n_route.py:9]: I want openroad to be called within this function. 
And maybe renambe this function to detailed place_n_route

File: openroad_interface/length_calculations.py [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+    '''
+    calculates lengths of each net using the def file
+
+    params: 
+        units: to convert length into correct units
+    returns: 
+        length_dict: length with macro ID attribution
+    '''
+
+    def_file = "results/final_generated-tcl.def"
```
Comment (fazerlicourice7) [inline openroad_interface/length_calculations.py:15]: this shouldn't be hardcoded. IT should instead be a parameter with a default value.

`def length_calculations(units: float, def_file: str = "results/final_generated-tcl.def") -> dict`

Comment (fazerlicourice7) [reply openroad_interface/length_calculations.py:15]: make this change in all the functions

File: openroad_interface/macro_maker.py [hunk 22]
```diff
@@ -0,0 +1,148 @@
+import math
+import copy
+import yaml
+
+from functions import *
+
+lef_tech_file = "OpenROAD/test/Nangate45/Nangate45_tech.lef"
```
Comment (fazerlicourice7) [inline openroad_interface/macro_maker.py:7]: this should not be hardcoded. 
We want this script to work for any specified lef file

File: openroad_interface/macro_maker.py [hunk 23]
```diff
@@ -0,0 +1,148 @@
...
+lef_tech_file = "OpenROAD/test/Nangate45/Nangate45_tech.lef" 
+spacing = None
+manufacturing_grid = None
+pin_size = 1000
+VDD = 1000
+VSS = 1000
+row_height = None
+spacing_constant = 3 
+die_coord_x2 = 100.13 * 1000
+die_coord_y2 = 100.8 * 1000
```
Comment (fazerlicourice7) [inline openroad_interface/macro_maker.py:16]: so, we specify a .lef and these should be extracted from that.

File: openroad_interface/graph_plotter.py [hunk 24]
```diff
@@ -0,0 +1,44 @@
...
+    '''
+    
+    box_plot = None
+    rcl = pd.read_csv('results/result_rcl.csv')
+    rcl[type_data] = np.log2(rcl[type_data])
+
+    sns.set_theme(rc={'figure.figsize':(10, 15)}, style="ticks")
+
+    flierprops = dict(marker='o', markerfacecolor='None', markersize=10,  markeredgecolor='black')
+    box_plot = sns.boxplot(data=rcl, x="design", y=type_data, hue="method", palette=["red", "blue"], width=0.5, showfliers=show_flier, flierprops=flierprops)
```
Comment (fazerlicourice7) [inline openroad_interface/graph_plotter.py:22]: pro tip.
Can just do 
plt.yscale('log') 
instead of manually converting the data to log.
and with this plt does some nice formatting stuff as well.

File: openroad_interface/graph_plotter.py [hunk 25]
```diff
@@ -0,0 +1,44 @@
...
+import matplotlib.patches as mpatches
+import matplotlib
+import numpy as np
+import pandas as pd
+import seaborn as sns
+
+def box_whiskers_plot_design(type_data: str, designs: list, units: str, title: str, show_flier: bool) -> None:
+    '''
+    generates boxplots
+    '''
```
Comment (fazerlicourice7) [inline openroad_interface/graph_plotter.py:13]: can we have better docs on this. 
What is `type_data`

File: openroad_interface/estimation.py [hunk 26]
```diff
@@ -0,0 +1,120 @@
...
+    edge_list = []
+    for x in range(1, len(net_list)):
+        edge = math.sqrt(pow(coord_data[the_component]["x"]/unit - coord_data[net_list[x]]["x"]/unit, 2) + pow(coord_data[the_component]["y"]/unit - coord_data[net_list[x]]["y"]/unit, 2))
+        edge_list.append(edge)
+        result += edge
+    return result, edge_list
+
+def global_estimation():
+    wire_global_file = directory + "results/wire_length_global.txt"
+    global_pattern = r"grt:\s_[0-9]+_\s[0-9]*\.[0-9]+\s[0-9]+"
```
Comment (fazerlicourice7) [inline openroad_interface/estimation.py:26]: should be params with default values

File: openroad_interface/macro_maker.py [hunk 27]
```diff
@@ -0,0 +1,148 @@
...
+    pins.append(pin)
+
+    pin ="   PIN VSS\n    DIRECTION INOUT ;\n    USE POWER ;\n    PORT\n      LAYER metal1 ;\n        RECT {} {} {} {} ;\n    END\n   END VSS\n".format( 0, (design_list["y"]-VDD)/1000, design_list["x"]/1000, design_list["y"]/1000, pin_name)
+    pins.append(pin)
+
+    return (pins)
+
+# specify how many rows and columns of pins you want ,and generates the design 
+def macro_maker(row_height, macro_list, op, heigh_num, pin_num_x, pin_num_y):
+    row_height *= heigh_num
```
Comment (fazerlicourice7) [inline openroad_interface/macro_maker.py:86]: docs

File: openroad_interface/macro_maker.py [hunk 28]
```diff
@@ -0,0 +1,148 @@
...
+        macro_x  = constant_macro_x - pin_size * pin_num_x
+        spacing_x = 0
+        if (spacing <= macro_x/(pin_num_x + 1)):
+            spacing_x = math.floor(macro_x/(pin_num_x + 1) / manufacturing_grid) * manufacturing_grid
+            macro_design = {"name": op + str(pin_num_x * pin_num_y) + "_" + str(size), "pin_x" : pin_num_x, "pin_y" : pin_num_y, "spacing_x" : spacing_x, "spacing_y" :spacing_y, "x" :constant_macro_x, "y" :row_height}
+            macro_list.append(macro_design)
+            # print ("invalid design")
+
+tech_params = yaml.load(open("../src/params/tech_params.yaml", "r"), Loader=yaml.Loader)
+area_list = tech_params["area"]
```
Comment (fazerlicourice7) [inline openroad_interface/macro_maker.py:109]: use closest technode to this specified lef

File: openroad_interface/macro_maker.py [hunk 29]
```diff
@@ -0,0 +1,148 @@
...
+######################################################################
+#######                                                        #######
+#######          all units are converted to nanometer          #######
+#######                x 1000 is the conversion                #######
+#######                                                        #######
+######################################################################
+
+# this script generates txt files that contain lef documentation for different generated macros
+
+def best(list):
```
Comment (fazerlicourice7) [inline openroad_interface/macro_maker.py:27]: make general script that takes lef file and list of operators as inputs. use Arg parser.
chose appropriate key in tech_params based on lef choice.
get aspect ratio closest to Golden aspect ratio (set in vars).
let rowheights vary from 1 - 34. 
out put the macro text for the chosen aspect ratio. 
this is the last thing to do. 
(maybe we can do this in a separate PR).

**Review by fazerlicourice7 — COMMENTED** [review]
.

File: openroad_interface/estimation.py (hunk 19 above)
Comment (daheehee) [reply openroad_interface/estimation.py:38]: i completed this, but doing this requires me to pass another argument to the estimation.

File: openroad_interface/exec.py [hunk 30]
```diff
@@ -0,0 +1,82 @@
+import shutil
```
Comment (fazerlicourice7) [inline openroad_interface/exec.py:1]: this file should be called validaiton_script or something similar

File: openroad_interface/README.md [hunk 31]
```diff
@@ -0,0 +1,17 @@
+# Openroad Interface
+
+Allows codesign to utilize OpenROAD for resistance, capacitance, and length calculations. The main function takes in two parameters, the graph file name and tcl file directory, and will return a pandas DataFrame containing all information about openroad rcl and estimated rcl.
+
+## Installation
+1. Build OpenROAD using the documentation that they provide: https://openroad.readthedocs.io/en/latest/user/Build.html
+2. Once OpenROAD is built, the command "openroad", which activates the tcl shell, should be able to run everywhere. If this is not the case, just ensure that this command is able to run in the OpenROAD test directory by exporting the route. To do this, go to the OpenROAD test directory (OpenROAD/test), then run:
+```
+export PATH=./../build/src:$PATH
+```
```
Comment (fazerlicourice7) [inline openroad_interface/README.md:10]: Change this to be the absolute path to openroad 
you can frame in terms of <path_to_your_codesign_root>/openroad_interface/OpenROAD/build/src
or whatever it should be.

File: src/simulate.py [hunk 32]
```diff
@@ -457,6 +456,26 @@ def main(args):
...
+    should not matter how many inputs are on because all the same voltage. t = rc, so that is simple. 
+    add them up probably. 
+    5. there was a third one i think, but i forgot what it was. 
+    6. the game plan is probably go through for the starting nodes and then work from there. check
+    if the node and its respective net and store that net in a set. then check the next node etc. then
+    using that net list, calculate respective information. then we go onto the next step in time. 
+    7. all of this will, probably, happen in this simulate function, but this will be checked. 
+    8. coding will be done once i can  GET THIS THING TO RUN. 
+    '''
+
```
Comment (fazerlicourice7) [inline src/simulate.py:478]: can you remove this comment here

File: src/benchmarks/models/matmult.py [hunk 33]
```diff
@@ -1,6 +1,7 @@
-from loop import loop
 import numpy as np
 
+from .loop import loop
```
Comment (fazerlicourice7) [inline src/benchmarks/models/matmult.py:3]: this import can actually be deleted.

**Review by fazerlicourice7 — COMMENTED** [review]
few tiny little cleanup things

File: openroad_interface/place_n_route.py [hunk 34]
```diff
@@ -0,0 +1,157 @@
...
+            input_nodes = list(graph_copy.in_edges(node))
+            output_node = list(graph_copy.out_edges(node))[0][1]
+
+            old_edge_1 = graph_copy[input_nodes[0][0]][node]
+            old_edge_2 = graph_copy[input_nodes[1][0]][node]
+
+            old_output = graph_copy[node][str(output_node)]
+            for attribute in ["net_length", "net_cap", "net_res"]:
+                old_edge_1[attribute] += float(old_output[attribute])
+                old_edge_2[attribute] += float(old_output[attribute])
```
Comment (fazerlicourice7) [inline openroad_interface/place_n_route.py:33]: I know this is what I originally asked for, but let's instead make it a list of all the Rs and Cs that contributed to this.

Comment (fazerlicourice7) [reply openroad_interface/place_n_route.py:33]: this well let us generalize our E and T calculation in the next step.

**Review by fazerlicourice7 — COMMENTED** [review]
.

**Review by fazerlicourice7 — APPROVED** [review]
alright. Let's send it. Hopefully you haven't broken anything.
